nncf/initialization.py

```diff
@@ -105,14 +89,25 @@ def write(cls, msg):
             initializers[name] = RangeInitializerFactory.create(init_config, module, log_module_name=name)
             hook_handles.append(module.register_forward_hook(initializers[name].forward_hook))
             module.init_stage = True
+
+        device = next(self.model.parameters()).device
+
+        data_loader = maybe_wrap_dataloader_for_init(data_loader)
         with torch.no_grad():
             bar_format = '{l_bar}{bar} |{n_fmt}/{total_fmt} [{elapsed}<{remaining}]'
             bar_desc = 'Algorithm initialization'
-            for i, (input_, _, dataloader_kwargs) in tqdm(enumerate(wrapped_data_loader), total=num_init_steps,
-                                                          desc=bar_desc, bar_format=bar_format):
+            for i, args_and_kwargs_tuple in tqdm(enumerate(data_loader), total=num_init_steps,
+                                                 desc=bar_desc, bar_format=bar_format):
                 if num_init_steps is not None and i >= num_init_steps:
                     break
-                self.model(input_, **dataloader_kwargs)
+
+                def is_tensor(obj):
+                    return isinstance(obj, torch.Tensor)
+
+                to_device_fn = partial(torch.Tensor.to, device=device)
+                args, kwargs = objwalk(args_and_kwargs_tuple, is_tensor, to_device_fn)
```

Comment:
the same changes are needed in hessian_estimator.

Reply:
Done